### ISSUE TITLE
Add support for [ObservableProperty] on partial properties

### ISSUE THREAD

**Issue by Sergio0694** [CommunityToolkit/dotnet#555]
### Overview

This issue tracks adding support for partial properties to the `[ObservableProperty]` generator. This is blocked on https://github.com/dotnet/csharplang/issues/6420, so support for this will have to wait for that first (hopefully in the C# 12 timeframe).

### API breakdown

The changes are pretty straightforward: properties will also be allowed on `[ObservableProperty]`:

```diff
-[AttributeUsage(AttributeTargets.Field, AllowMultiple = false, Inherited = false)]
+[AttributeUsage(AttributeTargets.Field | AttributeTargets.Property, AllowMultiple = false, Inherited = false)]
 public sealed class ObservablePropertyAttribute : Attribute
 {
 }
```

Analyzers will also need to be added to ensure that the annotated properties are partial and valid.

### Usage example

Code like this:

```csharp
[ObservableProperty] 
public partial string Name { get; private set; }
```

Will generate the following:

```csharp
/// <inheritdoc cref="_itemsSource"/>
[global::System.CodeDom.Compiler.GeneratedCode("CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator", "8.1.0.0")]
[global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage]
public partial string Name
{
    get => field;
    set
    {
        if (!global::System.Collections.Generic.EqualityComparer<string>.Default.Equals(field, value))
        {
            OnNameChanging(value);
            OnPropertyChanging(global::CommunityToolkit.Mvvm.ComponentModel.__Internals.__KnownINotifyPropertyChangingArgs.Name);
            
            field = value;

            OnItemsSourceChanged(value);
            OnNameChanged(global::CommunityToolkit.Mvvm.ComponentModel.__Internals.__KnownINotifyPropertyChangedArgs.Name);
        }
    }
}

/// <summary>Executes the logic for when <see cref="Name"/> is changing.</summary>
[global::System.CodeDom.Compiler.GeneratedCode("CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator", "8.1.0.0")]
partial void OnNameChanging(string value);

/// <summary>Executes the logic for when <see cref="Name"/> just changed.</summary>
[global::System.CodeDom.Compiler.GeneratedCode("CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator", "8.1.0.0")]
partial void OnNameChanged(string value);
```

This will also enable support for declaring custom accessibility modifiers for property accessors.

### Progress tracking

- [X] Initial code generation support
- [X] Add analyzer for C# preview
- [X] Add analyzer to use partial properties instead of fields
- [x] Add code fixer to convert fields to partial properties
- [x] Add analyzer for partial property declarations
- [x] Add analyzer for unsupported Roslyn version for properties
- [x] Update all titles/descriptions of existing shared diagnostics

**Comment by cmiles** [comment]
@Sergio0694 I haven't tried to track a language proposal before and couldn't immediately tell from the Partial Properties proposal/labels/information whether it is still expected to be available in the November 2023 time frame that you mentioned in #291 - since you have been tracking and adding to the the Partial Properties proposal/discussion any chance you can add any current information about time frame estimates here?

Fwiw From what I've read here and in other issues Partial Properties seem like a great solution - just wondering if it is truly 'coming soon' or if this pushes into 2024 and beyond.

Thanks for your time and any update!

**Comment by Sergio0694** [comment]
Hey @cmiles! We don't have any updates/ETAs to share at this time. Keep in mind that no C# features at all are confirmed until the moment they actually ship, so the same applies here. We're still actively looking into this though, just no news just yet 🙂

**Comment by modmynitro** [comment]
Its not in C# 12
https://github.com/dotnet/csharplang/issues/6420#issuecomment-1745771890

**Comment by cmiles** [comment]
@Sergio0694 since this did not make it into C#12 any chance that of any of the associated issues will be reconsidered for action in a more immediate time frame?

I understand partial properties would facilitate the ideal implementation of things like 'Required' keyword support - but there is some functionality blocked by this that would be immediately useful and appreciated even if there were breaking changes in the future if/when partial properties make it into C#.

**Comment by Sergio0694** [comment]
> "but there is some functionality blocked by this"

What functionality is blocked?

**Comment by cmiles** [comment]
> What functionality is blocked?

@Sergio0694 - It was my understanding that all of the issues/functionality listed below are blocked by wanting to wait and create an implementation based around the partial property language feature.

For me the functionality the required keyword offers is particularly attractive as we work on enabling nullable support in our code - but I'm not aware of any way to combine [ObservableProperty] and required?

Based on your question of 'What functionality is blocked?' maybe I am mistaken and there is a feature or approach that I'm not aware of to use [ObservableProperty] and required together - if so apologies for the wasted time/space in this issue but would very much appreciate a link to get me started!

[Add support for "required" and "virtual" keywords to property generator · Issue #679 · CommunityToolkit/dotnet](https://github.com/CommunityToolkit/dotnet/issues/679)
[Make ObservableProperty Generated Code Properties Virtual? · Issue #543 · CommunityToolkit/dotnet](https://github.com/CommunityToolkit/dotnet/issues/543)
[Add support for setter protection level for ObservableProperty · Issue #758 · CommunityToolkit/dotnet](https://github.com/CommunityToolkit/dotnet/issues/758)
[Add abilities to control the access level of generated [ObservableProperty] attribute · Issue #585 · CommunityToolkit/dotnet](https://github.com/CommunityToolkit/dotnet/issues/585)

**Comment by Sergio0694** [comment]
> "all of the issues/functionality listed below are blocked"

> "For me the functionality the required keyword offers is particularly attractive"

No functionality is actually blocked. You can easily use all that — just manually define the properties 🙂

I understand the frustration for wanting a simpler way to do that, and I do agree that of course using the generator is better. But I don't think it's a good idea to dedicate resources to implement all that functionality when we might very well have partial properties out in preview soon, especially because then we'd still have to keep supporting that "workaround" to avoid breaking consumers. Rather, it makes more sense to just wait for partial properties to be available in preview, add support for those, and then release a preview of the MVVM Toolkit too so that people wanting to try those out can do so as well.

**Comment by cmiles** [comment]
@Sergio0694 No frustration here - seems like we just have a different evaluation of the issue.

Fwiw I would still '+1' working on a solution now - but respect and understand your reasoning and decision and no doubt you have 100% better insight on the odds of partial properties arriving in a near future preview. 

I hope this issue will continue to be re-evaluated if for some reason partial properties, or something equivalently useful for generators, don't arrive for the next version of C#.

Thanks for the prompt replies - appreciated!

**Comment by jhm-ciberman** [comment]
This was not added to C# 12. Is there any reason to not implement the other more traditional alternatives using attributes?

**Comment by Sergio0694** [comment]
This exact question and the answer for that are in the messages right above yours — the reason has not changed and it's the same as before 🙂

**Comment by AignerGames** [comment]
Partial properties are just in proposel state, it's not confirmed that they will be release any time soon (or at all). I don't think that it is a good idea to wait for some unknown future time and close all releated issues. 

I understand your argument, that the current solution is a workaround, but it's a workaround that already exists, so adding more features to the already existing / working "workaround" would be better for every current user of the toolkit. You can declare it obsolete as soon as partial properties are implemented (if ever) and then you don't have to worry about maintaning it in the future. 

Would it really take that many resources to add a additional parameter to control the accessor kind of the generated property to justify waiting for partial properties?

**Comment by DennisWelu** [comment]
@AignerGames  is right on. I’ve had similar thoughts in the past about this issue. It could be years or never, yet the value is high and the effort seems at least relatively low. Worth the trouble of a depreciation period if needed.

**Comment by jhm-ciberman** [comment]
Personally, I don't use generators in MVVM Toolkit at all only for this sole issue.

**Comment by jhorv** [comment]
@Sergio0694 frustration here.

The fact that `[ObservableProperty]`s are public-public causes me to barely use it. If only I could specify a `private set`, then almost every one of my view models could have its lines of code cut in half. What's the point of deriving from `ObservableObject` if I'm always the one calling `SetProperty()`? I could have that with an extension method, for crying out loud.

It's becoming increasingly difficult to convince people (and believe myself) that I did a smart thing in choosing this library. To me it doesn't matter how, whether via attribute or some future C# language feature. Just add the ability and make these features useful for more than just demos.

**Comment by Sergio0694** [comment]
I understand the frustration, however how I've already said multiple times in several related issues, my goal is to structure the Toolkit in a way that ensures the best path forward for it and for users, and I don't think trying yo shoehorn additional APIs to an already arguably "hacky" API solely for a temporary workaround is the right solution. We have millions of users and if we rolled that out, it's also something we'd likely have to maintain for at least some time, as we wouldn't want to just break everyone by immediately obsoleting and removing that API once we did get partial properties.

You can also check the [Roslyn feature status](https://github.com/dotnet/roslyn/blob/main/docs/Language%20Feature%20Status.md) and you will see that partial properties have now been added to the working set. I'll prototype support for the MVVM Toolkit and publish a preview version with them as soon as they're available in a preview .NET SDK. Once again, this has always been the plan.

> "What's the point of deriving from ObservableObject if I'm always the one calling SetProperty()? I could have that with an extension method, for crying out loud."

That is just incorrect. You wouldn't be able to raise an event from an extension method, only code within that class can raise events. This is just a C# limitation. Also I'm not even sure what you mean here, deriving from `ObservableObject` is not even required at all. It's just recommended to minimize binary size.

I will also say, as a side node and personal opinion, I don't really understand how not having support for private setters is a deal breaker. Yes, it's not as clean, but being pragmatic, it genuinely doesn't matter. Just ignore that, use the generators on fields, don't use the setters from the outside for now, and just go back and update the code once we get partial properties. This is exactly the same we're doing in the Microsoft Store app. The benefit of using the generators and saving hundreds and hundreds of lines of code vastly outweighs the fact that "oh this setter in this viewmodel is public but it should be private, even though literally nobody is setting from the outside anyway so it doesn't make any difference in practice". You're building an app, not a library, so you own the entire stack. Just don't use the setters for now, it will be fine.

**Comment by AignerGames** [comment]
I don't see the high maintenance or fear of breaking the feature in the future as valid, because the C# syntax will not suddenly change so drastically that a setter modifier will be handled totally different or handled differently all the time. 

But after reading the comments and other issues it's very clear that you will not allow it no matter what, so I will give up and not argue any further.

**Comment by weitzhandler** [comment]
Having private/protected/virtual setters or changing/changed methods is essential. Please add properties to the Observable property attribute to control that.

**Comment by Sergio0694** [comment]
> "I don't see the high maintenance or fear of breaking the feature in the future as valid, because the C# syntax will not suddenly change so drastically that a setter modifier will be handled totally different or handled differently all the time."

@AignerGames I think there might be a misunderstanding here. It's not the C# syntax that would change. It's the fact that if we released support for customizing setters on `[ObservableProperty]`, then people would start using it, then C# would get support for partial properties (which again, it's literally in the working set now), and once that shipped we'd add support for that. We'd then be in the situation where we'd either have to support both, which we don't want, or deprecate and remove that temporary support for customizing setters on fields, which would mean breaking all of those consumers that had started using that new feature. We don't want that. We want people to just switch to the right thing (partial properties), only have to change their code once, and never get broken by us deprecating that feature in the first place.

> "that you will not allow it no matter what, so I will give up"

It's not like I will not allow it just because, I just want to do things right from the start, especially now that partial properties are finally in the working set and might be available soon, is all. I do want this feature to be available to everyone, and I also do want it myself 🙂

> "Having private/protected/virtual setters or changing/changed methods is essential. Please add properties to the Observable property attribute to control that."

@weitzhandler it is certainly a nice to have and an important feature, but it is not essential (or there wouldn't be several production apps using it despite that lack of support for now). As we mentioned in the rest of this thread, we have no plans on adding that feature _for fields_. We plan on supporting that for partial properties as soon as they're added to C#, however.

**Comment by todor-dk** [comment]
@Sergio0694 I am with you. I really want private, protected etc., but keep the implementation clean. I am willing to wait 6 months if that's what it takes.

On why I want private / protected / etc.; The answer is easy: **maintainability** of my code. If it is private, when refactoring code, I know there won't be anybody outside of my class fiddling with the property - less headache when refactoring a class. Same goes for the other modifiers. And this is especially important if the property is a critical one.

Second, a thing I hope something partial properties will solve, which the current infrastructure cannot solve, is the ability to **find references** in the IDE. Today, if I have an observable property, I cannot easily take references, as if I try on the field, let's say on **_Name**, I only get the auto-generated setter, and then I have to go one level further and find references on **Name**. This is just inconvenient, and if partial properties will solve this, I am willing to wait few months.

**Comment by DennisWelu** [comment]
Requests for this capability in this library go back to 2022 from what I can see. AFAICT, the C# feature we are waiting on (https://github.com/dotnet/csharplang/issues/6420) is still in the proposal stage at this point? I'm not seeing any indication it will be included this fall. Could be missing something there. If that's true, I think people have been holding out for "a few months" for quite some time now. Am I correct in assuming that if later this feature is added to the language then related changes in ObservableProperty will largely be in the generated code? Seems like minimal impact on the usage protocol for ObservableProperty itself.

High value in this feature. Worth a deprecation period if necessary.

**Comment by jhm-ciberman** [comment]
If you don't want to keep supporting a feature in the long run, can we just add it with a message in the documentation that says something like:

"This feature will be deprecated and removed once partial properties are added to the C# language. If you use this feature, you acknowledge this."

**Comment by jphorv-bdo** [comment]
Why deprecate the attribute approach at all? It is already a compile error to add `[ObservableProperty]` in a type that isn't `ObservableObject`/`[ObservableObject]`/`[INotifyPropertyChanged]`. If access level can be specified via attribute, it could simply be a compile error to specify it on a partial property. There are companies who don't move to the latest/greatest versions immediately, and there will be people who would rather specify their observable properties via a field than a partial property.

**Comment by Sergio0694** [comment]
> "is still in the proposal stage at this point?"

@DennisWelu like I mentioned, it's in the working set. [Here](https://github.com/dotnet/roslyn/issues/73090) you can see the test plan (and the VS 17.12 milestone) and a link to the feature branch. You can see work has started and there's a couple of open PRs starting to add support for it 🙂

> "If you don't want to keep supporting a feature in the long run, can we just add it with a message in the documentation that says something like"

I don't want to spend weeks implementing the feature just to deprecate it shortly after.

> "Why deprecate the attribute approach at all?"

Because I don't want to keep having to maintain two parallel implementations in the source generators. We might keep support for the current attribute on fields for longer, but I don't want to _also_ add all the additional complexity for custom accessors, and having to maintain that too for potentially a long period of time. Like I said, I want to just get this right from the start.

**Comment by DennisWelu** [comment]
@Sergio0694 Thanks for the pointer to the ongoing work! Hopefully since work has started it will make the final cut. Seems hopeful...

Also appreciate your efforts to keep the API clean and keep efforts focused/manageable. Good things for sure.

My understanding is you would not maintain the Attribute approach after this feature is available. Which means consumers are going to want to update code en masse anyway at some point? That's fine by me. Changing over might be tedious but it should be a clear-cut recipe.

Still...probably I don't understand the "weeks" of implementation needed but I imagine adding something in the meantime as simple as `SetterVisibility=Visibility.Private` as an option to ObservableProperty, and all that does is change the visibility keyword used in the generated code. FWIW.

This will continue to be important to our team because we are porting a bunch of Xamarin projects over to MAUI and the one big change is incorporating the MVVM toolkit where possible. The choice today for read-only properties is to make them read-write, or don't adopt generated properties. But there's SO much goodness in them it's hard not to. It will be hard to find all these properties later because nothing discretely points them out if you go looking for them.

**Comment by AignerGames** [comment]
> Still...probably I don't understand the "weeks" of implementation needed but I imagine adding something in the meantime as simple as `SetterVisibility=Visibility.Private` as an option to ObservableProperty, and all that does is change the visibility keyword used in the generated code. FWIW.
> 


I don't have experience with this code base, so my information could be wrong, but after a quick check I think that only a single place of the generator would have to be changed [here](https://github.com/CommunityToolkit/dotnet/blob/7b53ae23dfc6a7fb12d0fc058b89b6e948f48448/src/CommunityToolkit.Mvvm.SourceGenerators/ComponentModel/ObservablePropertyGenerator.Execute.cs#L1003) to add the modifier type. 

Well and the code to define the enum somewhere, modify the attribute to accept the enum and maybe unit tests.

**Comment by jhorv** [comment]
> I will also say, as a side node and personal opinion, I don't really understand how not having support for private setters is a deal breaker. Yes, it's not as clean, but being pragmatic, it genuinely doesn't matter. Just ignore that, use the generators on fields, don't use the setters from the outside for now, and just go back and update the code once we get partial properties. This is exactly the same we're doing in the Microsoft Store app. The benefit of using the generators and saving hundreds and hundreds of lines of code vastly outweighs the fact that "oh this setter in this viewmodel is public but it should be private, even though literally nobody is setting from the outside anyway so it doesn't make any difference in practice". You're building an app, not a library, so you own the entire stack. Just don't use the setters for now, it will be fine.

For my organization and project, it matters **a lot**. The project I'm working on isn't just an app, but also a collection of libraries that other internal development teams will use. There is a client API library, a view model library, and a UI component library, and other devs may use all of it or just parts. As the provider of these libraries, it is a **very big deal** that these libraries be hard to use incorrectly. If Intellisense shows that something can be done/set/called then it must be a valid operation. That is the point of object-oriented encapsulation/data hiding, and all these properties with `public` setters is the opposite of that: a nightmare for supportability, dev productivity, and our project's success.

**Comment by Sergio0694** [comment]
Using an attribute is strictly worse than using partial properties.

**Comment by AignerGames** [comment]
I think we all agree that partial properties would be the best solution, but they don't exist yet and will not be released anytime soon.

**Comment by Sergio0694** [comment]
They are quite literally already merged in the Roslyn repo and will be available in VS 17.11 Preview 3, out shortly 😄
With them being part of C# 13 pretty much for sure at this point, I'm not going to be implementing a workaround for this.

**Comment by AignerGames** [comment]
Yes the **preview** version, but most people will not use a preview version. The next full release is currently planned for November 24 (if no delays happen) so almost 2 years after this issue was opend. I agree that we can wait now, but adding a "workaround" 2 years ago would still have been better, since all the closed requests over the years prove that there was a use case for it.

**Comment by Sergio0694** [comment]
Hindsight is 20/20. Partial properties were pushed back multiple times, we didn't know it'd take this long at the time 🙂

**Comment by AignerGames** [comment]
I guess you are more optimistic than me, I don't trust any of their release dates until the feature is implemented, because I waited for many C# features in the past that were pushed back or only proposed and then never worked on until this day 😂

**Comment by pos777** [comment]
> Using an attribute is strictly worse than using partial properties.

Agreed. If you can use partial properties, then you should do so =) But they are not always available. And I do not mean that they are not implemented yet. For example, the codebase may be pinned to C# 11 for some reason.

As stated in the https://github.com/dotnet/csharplang/blob/main/proposals/partial-properties.md article, the programmer must specify the access modifier when implementing partial properties:
```cs
partial class C1
{
    public partial string Prop { get; private set; }
    public partial string Prop { get => field; private set => field = value; }
}
```
Thus, the logic for generating the access modifier is required in this case too. In the case of partial properties, the information about required access modifier is available in the defining declaration. In turn, in the case of using fields, similar information could be taken from a special attribute. So we get
- if it is possible to use partial properties:
```
[ObservableProperty]
public partial string Name { get; private set; }
```
- if the use is limited to fields:
```
[ObservableProperty, WithPrivateSetter]
private string name;
```
If I understood correctly from the discussion, you do not want to modify the `ObservableProperty` attribute so as not to confuse users after the partial properties appear. Storing information in a separate attribute solves this problem, in my opinion. I will prepare a request to demonstrate the approach.

However, all this makes sense if you do not plan to abandon the approach with fields.

**Comment by LE-MUC-MarkusWolff** [comment]
Hi! I just came across this issue as I was looking for a solution to a problem that might be solved by this. I would definitely appreciate if a feature like this were available. I also just found out that the [C# updates in .NET 9 Preview 6](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/csharp.md#partial-properties) actually includes the partial properties! 

A first step might be for the `[ObservableProperty]` to generate a partial property, which to my knowledge should not break anything, while preserving the field approach. I am not sure if this will solve all of my use cases, but I would at least give it a try.

**Comment by Sergio0694** [comment]
Generating a partial property wouldn't really do anything, other than making your code fail to compile 🙂

As for the real support in the MVVM Toolkit, I'm waiting to see whether `field` will make it as well.

**Comment by LE-MUC-MarkusWolff** [comment]
Ok, then my knowledge is not reaching far enough. I hope something like this can be implemented though.
I would need to annotate the generated Properties with Attributes for EF Core. But since the Navigation Properties have to be virtual anyway, I will probably have to o it by hand for the foreseeable Future anyway.

**Comment by LE-MarkusW** [comment]
> Generating a partial property wouldn't really do anything, other than making your code fail to compile 🙂

Ok I played around with the Preview a little bit. I guess the Introduction of `[PartialProperty]` and `[VirtualProperty]` would probably be necessary for my use case. :(

**Comment by cmiles** [comment]
@LE-MarkusW I don't really know your use case but if you haven't already it might be worth your time to look at [Metalama](https://www.postsharp.net/metalama) for code generation. Metalama is a paid tool with a free tier - [Pricing](https://www.postsharp.net/metalama/pricing), [License Information](https://github.com/postsharp/Metalama).

I used it to write my own INotifyPropertyChanged a little over a year ago in order to take advantage of the required keyword and  and now use it for a few other generation tasks including Commands. For my simple requirements getting to working code was pretty quick.

Metalama isn't my focus - you can see much more beautiful code in the Metalama documentation - but here is my current code [PointlessWaymarksTools/PointlessWaymarks.LlamaAspects)).

**Comment by hez2010** [comment]
It would be better to warn users on the old field usage as well and provide a code fix to help developers migrate from attributes on fields to attributes on partial properties after we ship the support.
The current way doesn't work well with COM/WinRT generators as all those properties are generated so that they cannot be seen by other source generators.

**Comment by Sergio0694** [comment]
I plan on adding a code fixer for that as well, yes 🙂

**Comment by heartacker** [comment]
hello, when can we get this with dotnet 9 released?

**Comment by AndrewKeepCoding** [comment]
> hello, when can we get this with dotnet 9 released?

- Install [CommunityToolkit.Mvvm v8.4.0-preview1](https://www.nuget.org/packages/CommunityToolkit.Mvvm/8.4.0-preview1).
  (Make sure you enabled the **Include prerelease** checkbox.)
- In your *.csproj file, set the language version to `preview`.
  (`field` didn't make it to C#13 but is expected to be on C#14.)
  ```xml
  <PropertyGroup>
    <LangVersion>preview</LangVersion>
  </PropertyGroup>
  ```

**Comment by Mrxx99** [comment]
When is a stable version released with this?

**Comment by weitzhandler** [comment]
Hi,

I've been using backed field properties for a while now and it works amazing, thank you!

There's one thing in particular I'd like to share.

Since there's no way to access the field from outside the property, a change notification is always going to be triggered, even when setting the property from the constructor.
This is totally redundant in most cases.

Is there a way to suppress INPC or set the backing field directly when setting the property in constructor?

**Comment by SirViver** [comment]
Actually, this is quite a big issue IMO. 

The change to partial properties potentially causes breaking changes when converting an existing codebase, as there might be a ton of `override OnPropertyChanged(...)` or `partial On{Property}Changed(...)` that simply break behavior or cause exceptions when executed from within the constructor or property initializer.

Is there way other than adding a `bool _isInitialized` field to every class that is set at the end of the ctor and then manually checked in each and every `OnXXXChanged` to avoid premature property change actions?

Maybe provide an `protected Initialize(Action initializer)` that internally suppresses any property changed notifications/calls, and the user has to change the ctor to somthing like this:
```
public class MyClass : ObservableObject
{
    [ObsevableProperty]
    public partial bool SomeProp { get; set; }

    public MyClass()
    {
        Initialize(() => 
        {
            SomeProp = true;
        });
    }
}
```

Not exactly sexy (and doesn't allow for property initializers at all), but I don't think there's really a native way to detect "derived ctor has finished executing" in C# ☹ 

I mean, ultimately it will simply not be possible to just convert everything to partial properties - sometimes you just need the possibility to change a property value without triggering change notifications - so living in a mixed mode world where some properties use fields and some use partial properties is probably unavoidable, but maybe there ought to be a bit more of a warning about the potential pitfalls when/before executing the suggested CodeFix conversion to partial properties.

**Comment by jphorv-bdo** [comment]
I also have unexpected issues with this partial property feature, and I'm unsure I'll be able to use it nearly as much as I'd hoped. Notifications being sent while still in the constructor causes a lot of issues in my codebase. What started with me happily deleting hand-written properties that existed only because I wanted private setters, ended with me stashing the entire changeset.

**Comment by SirViver** [comment]
Another (ugly/hack) workaround would be to make a `SetPropertyWithoutNotification` method that directly sets the property backing field via reflection.

```csharp
// Custom ObservableObject-derived base class defines this method:

protected void SetPropertyWithoutNotification<T>(T property, T newValue,
    [CallerArgumentExpression(nameof(property))] string propertyName = "")
{
    if (propertyName.LastIndexOf('.') is int dotIndex and >= 0)
    {
        propertyName = propertyName[(dotIndex + 1)..];
    }

    string backingFieldName = $"<{propertyName}>k__BackingField";
    
    var backingField = this.GetType().GetField(backingFieldName, BindingFlags.Instance | BindingFlags.NonPublic) 
        ?? throw new InvalidOperationException($"Backing field '{backingFieldName}' not found for property '{propertyName}'.");
    
    backingField.SetValue(this, newValue);
}

// Use in derived class like so:

[ObservableProperty]
public partial int MyProperty { get; set; }

public MyClass()
{
    SetPropertyWithoutNotification(MyProperty, 42);
}
```
Of course the usual caveats regarding reflection performance and ignoring/circumvention of property access modifiers apply, but... 🤷

**Comment by Sergio0694** [comment]
@weitzhandler @SirViver we're looking at fixing this in C# 14 🙂
See: https://github.com/dotnet/csharplang/issues/9031.

**Comment by AignerGames** [comment]
The fieldof thing looks weird, I think the easier and more obvious solution would be to give more control the the source generator. Instead of always using the field keyword for the generated property, add a parameter to the property attribute to generate a "real" field with a name. Then you could just write the value to the generated field like before, but still use the new features which are only possible with partial properties.  

Would be a clean and easy solution, without waiting for a new language feature, which is in proposal state, so it's not confirmed that it will be added in C# 14 or at all.

**Comment by weitzhandler** [comment]
Another method is adding a `IDisposable SuppressChangeNotifications()` method in a fashion similar to what we have in the `ObservableCollection`:

```csharp
class MyClass : INPC
{
    [ObservableProperty]
    partial int MyProperty { get; set; }

    public MyClass
    {
        using (SuppressPropertyChangeNotifications())
        {
            MyProperty = 5;
        }
    }
}
```

**Comment by DennisWelu** [comment]
@Sergio0694 I appreciate the entire body of work done in this library it is quite nice. But the philosophy of waiting for the next C# version continues to undermine the usefulness of the ObservableProperty, which may be the most used feature. I am hoping you will reconsider a solution for "today" in the meantime.

**Comment by JochemPalmsens** [comment]
> [@Sergio0694](https://github.com/Sergio0694) I appreciate the entire body of work done in this library it is quite nice. But the philosophy of waiting for the next C# version continues to undermine the usefulness of the ObservableProperty, which may be the most used feature. I am hoping you will reconsider a solution for "today" in the meantime.

IMHO If you think this is a problem, don't use it. Make your own: you can even clone this repo and modify it yourself.
Or just be glad that they're offering this toolkit at all, which saves us a lot of work and makes code more readible. And that they are actually reading and addressing our issues at al.
At least, that's how I'm experiencing it. I'm happy with this new feature and glad with the partial property support in .net 9, so we can make stuff `required`. Thanks guys!

**Comment by DennisWelu** [comment]
@JochemPalmsens I led with a statement of appreciation, which is genuine. And everything you say is valid. But it makes no sense to fork and maintain a large repo every time you disagree with a design or decision. And since this is a community project I assume it is open to community opinions. The history on waiting for a key change in the C# spec on this specific feature, the most valuable feature IMO, goes back years.

**Comment by AignerGames** [comment]
> Another method is adding a `IDisposable SuppressChangeNotifications()` method in a fashion similar to what we have in the `ObservableCollection` - calling the method disables notification, disposing its result enables it back on.
> 
> class MyClass : INPC
> {
>     [ObservableProperty]
>     partial int MyProperty { get; set; }
> 
>     public MyClass()
>     {
>         using (SuppressPropertyChangeNotifications())
>         {
>             MyProperty = 5;
>         }
>     }
> }
> 
> This is a useful utility method which comes handy regardless of partial properties or initialization.

I like this idea and it could even be combined with the other ideas, if desired.

**Comment by whiskhub** [comment]
I have had created a ticket for this feature request already: https://github.com/CommunityToolkit/dotnet/issues/993

After discussions I have then created the C# language request, but it seems they will not add it to the language anytime soon as they don't see the justification for a language level fix: https://github.com/dotnet/csharplang/discussions/8704#discussioncomment-12085677

Therefore I'd also really like to see a feature improvement in the source generator. An `IDisposable` wrapper (https://github.com/CommunityToolkit/dotnet/issues/555#issuecomment-2607412514) seems sensible to me.

**Comment by mbencew** [comment]
Shouldn't we be able to use this feature with C# 14 (instead of preview)? Or do we have to wait for a new version?

**Comment by joehanna** [comment]
Hey @Sergio0694 - I arrived here from [your comment](https://github.com/CommunityToolkit/dotnet/issues/835#issuecomment-2270171602) on [Issue #835](https://github.com/CommunityToolkit/dotnet/issues/835) which says that issue is superseded by this one. I am not clear how, but I didn't want to create yet another issue and I am not certain this comment belongs in this one. All of the Issues on my topic are linked forward to this one. I apologise if I have done this incorrectly.

I am still interested to know how to solve the need to override the generated public property or at least where I can vote for it. 

We don't use `lowerCamelCase` for our private fields so the `ObservableProperty` does not get generated as expected. Given coding style is personal and configurable using `.editorconfig` and most modern IDEs and Code Editors respect it, it makes it difficult consuming this library feature that is strictly opinionated on coding style. I am not expecting it to automatically support multiple styles, but having the ability to override the generated property name would solve that problem for everyone.

For example (as inspired by the example on the also-linked [Issue #795](https://github.com/CommunityToolkit/dotnet/issues/795):

```
[ObservableProperty("SpecialValue")]
private double value_using_WhatEVER_coding_style_I_makeup_oR_want;
```

For the record, we use `lower_snake_case` for private fields.

Would you kindly re-consider support for the public property name as an attribute? It is clean, readable, doesn't require tracking of partial properties and makes this feature accessible to us - otherwise we are off the proverbial cliff and back to full boilerplate property declarations :)

Thanks!
